smarts/core/vehicle_index.py

```diff
@@ -372,7 +347,17 @@ def build_agent_vehicle(
             initial_speed,
         )
 
-        waypoint_paths = sim.waypoints.waypoint_paths_at(vehicle.position, lookahead=1)
+        if vehicle_exists:
+            old_vehicle = self._vehicles[vehicle_id]
+            sensor_state = self._sensor_states[vehicle_id]
+            mission_planner = mission_planner or sensor_state.mission_planner
+            # Apply the physical values from the old vehicle chassis to the new one
+            vehicle.chassis.inherit_physical_values(old_vehicle.chassis)
+
+            # Remove the old vehicle
+            self.teardown_vehicles_by_vehicle_ids([vehicle_id])
+            # HACK: Directly remove the vehicle from the traffic provider
+            sim._traffic_sim.remove_traffic_vehicle(vehicle_id)
```

Comment:
We really need to avoid these hacks it breaks down the boundaries between classes. Instead, we should revise our interfaces as necessary to accomplish what was intended here. `VehicleIndex` is now directly talking to a provider. It should be going through the provider interface. At the very least, we could hub-and-spoke through SMARTS. But we certainly don't want to introduce touch points between the index and providers directly.

Reply:
This line is from:
https://github.com/huawei-noah/SMARTS/pull/48/files#diff-6113417ab844aea7f78d69bb51a162a0faecc6323fd5222cdaf55867386bc186L302

The main pain point here is that when trying to recreate is the timing of how the vehicles can be added. Currently it takes two steps for a provider added vehicle to get into the simulation and one step to delete a SUMO vehicle. It is also not possible to add a vehicle with the same id while another one exists. This is a problem because if trying to negotiate the problem through the provider it will take three steps total: two steps for the provider vehicle to reach the traffic sim, and then another step to delete the traffic vehicle. That extra delay has adverse effects elsewhere.

I think it might be possible to remove the `reserve_traffic_location_for_vehicle` call to the traffic sim but it perhaps needs more discussion as to if SMARTS should have the ability to call to remove a traffic vehicle.

Reply:
Want to talk about this in our sync tonight?